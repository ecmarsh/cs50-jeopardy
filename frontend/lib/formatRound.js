export default function(page) {
  let title;
  switch (page) {
    case 1:
      title = 'Single';
      break;
    case 2:
      title = 'Double';
      break;
    case 3:
      title = 'Final';
      break;
    default:
      title = 'CS50';
  }

  return title + ' Jeopardy';
}
