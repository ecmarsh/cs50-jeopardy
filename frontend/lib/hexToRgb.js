export default function(hex) {
  const chunk = 2,
    radix = 16,
    rgb = [];

  let i = 1;
  while (hex && i < hex.length) {
    rgb.push(parseInt(hex.substring(i, i + chunk), radix));
    i += chunk;
  }

  return rgb.toString();
}
