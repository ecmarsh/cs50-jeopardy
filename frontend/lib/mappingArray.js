export default function(N = 0, isFilled = true, startAtZero = false) {
  if (N == 0) return [];
  if (N > 0 && !isFilled) return Array(N).fill();
  if (isFilled && !startAtZero)
    return Array.from({ length: N }, (v, k) => k + 1);
  return Array.from({ length: N }, (v, k) => k++);
}
