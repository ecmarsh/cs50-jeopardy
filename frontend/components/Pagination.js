import React from 'react';
import PropTypes from 'prop-types';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import Link from 'next/link';
import formatRound from '../lib/formatRound';
import Error from './ErrorMessage';
import PaginationStyled from './styles/Paginated';
import { categoriesPerPage } from '../config';

const PAGINATION_QUERY = gql`
  query PAGINATION_QUERY($gameName: String) {
    categoriesConnection(where: { game: { name: $gameName } }) {
      aggregate {
        count
      }
    }
  }
`;

const Pagination = ({ gameName, round }) => {
  const labels = {
    prevPageLinkText: `Prev Round`,
    nextPageLinkText: `Next Round`,
    pathnameOfPage: `load`,
  };
  return (
    <Query query={PAGINATION_QUERY} variables={{ gameName: gameName }}>
      {({ data: { categoriesConnection }, loading, error }) => {
        if (loading) return <p>Loading...</p>;
        if (error) return <Error error={error} />;
        const totalPages = Math.ceil(
          categoriesConnection.aggregate.count / categoriesPerPage
        );
        return (
          <PaginationStyled data-test="pagination">
            <h4 data-test="pagination-round">{formatRound(round)}</h4>
            <Link
              prefetch
              href={{
                pathname: labels.pathnameOfPage,
                query: { game: gameName, round: round === 1 ? 1 : round - 1 },
              }}
            >
              <a className="prev" aria-disabled={round <= 1}>
                <span>←</span>
                {labels.prevPageLinkText}
              </a>
            </Link>
            <Link
              prefetch
              href={{
                pathname: labels.pathnameOfPage,
                query: { game: gameName, round: round + 1 },
              }}
            >
              <a className="next" aria-disabled={round >= totalPages}>
                {labels.nextPageLinkText}
                <span>→</span>
              </a>
            </Link>
          </PaginationStyled>
        );
      }}
    </Query>
  );
};
Pagination.propTypes = {
  gameName: PropTypes.string.isRequired,
  round: PropTypes.number.isRequired,
};

export default Pagination;
export { PAGINATION_QUERY };
