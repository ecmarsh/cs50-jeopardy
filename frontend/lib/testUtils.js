import casual from 'casual';
import mappingArray from './mappingArray';

// seed it for repeatable random sequence
casual.seed(123);

const fakeQuestion = () => ({
  id: 'abc123',
  question: 'Time complexity of merge sort?',
  answer: 'n log n',
  user: {
    id: casual.uuid,
    name: 'User Name',
    __typename: 'User',
  },
  createdAt: '2019-02 - 26T19: 24: 16.000Z',
  game: {
    id: 'game123',
    name: 'GameForGeekz',
    user: {
      id: casual.uuid,
      __typename: 'User',
    },
    isPublic: casual.coin_flip,
    __typename: 'Game',
  },
  categoryQuestions: [
    {
      id: 'catQuest123',
      category: { id: 'cat123', __typename: 'Category' },
      __typename: '__CategoryQuestion',
    },
  ],
  __typename: 'Question',
});

const fakeUser = () => ({
  __typename: 'User',
  id: 'userABC123',
  name: casual.name,
  email: casual.email,
  permissions: ['USER'],
});

const fakeAdmin = () => ({
  __typename: 'User',
  id: 'adminABC123',
  name: casual.name,
  email: casual.email,
  permission: ['ADMIN', 'USER'],
  games: [],
});

const fakeCategoryQuestion = (id = 'catQuest123', difficulty = 1) => ({
  id,
  difficulty,
  isDouble: false,
  answered: casual.coin_flip,
  question: fakeQuestion(),
  category: { id: 'cat123', name: casual.word, __typename: 'Category' },
  game: { id: 'g@m3', name: 'Test Game', __typename: 'Game' },
  __typename: 'CategoryQuestion',
});

const fillCategoryQuestions = (n = 5) =>
  mappingArray(n).map(i => fakeCategoryQuestion(`catQuest-${i}`, i));

const fakeCategory = (id = 'cat123') => ({
  id,
  name: casual.word,
  game: { id: 'g@m3', name: 'Test Game' },
  categoryQuestions: fillCategoryQuestions(),
  __typename: 'Category',
});

const fillCategories = (n = 11) =>
  mappingArray(n).map(i => fakeCategory(`cat-${i}`));

const fakeTeam = id => ({
  __typename: 'Team',
  id: `t${id}`,
  name: `Team${id}`,
  score: 1000,
  game: { id: 'g@m3', name: 'Test Game' },
});

const fakeGameConfig = () => ({
  id: 'config123',
  hasDoubleJeopardy: false,
  hasRoundTimer: false,
  roundTime: 0,
  finalTime: 0,
  __typename: 'GameConfig',
});

const fakeGame = () => ({
  __typename: 'Game',
  id: 'g@m3',
  name: 'Test Game',
  createdAt: '2019-03 - 06T19: 12: 16.000Z',
  isPublic: casual.coin_flip,
  user: fakeUser(),
  teams: [fakeTeam(1), fakeTeam(2)],
  categories: fillCategories(),
  questions: [fakeQuestion()],
  config: fakeGameConfig(),
});

const fakeStudyGame = (id = 'SG123') => ({
  __typename: 'StudyGame',
  id,
  answeredQuestions: [fakeCategoryQuestion(), fakeCategoryQuestion()],
  score: 1000,
  game: fakeGame(),
  user: fakeUser(),
  createdAt: '2019-04 - 25T19: 10: 16.000Z',
});

const fakeGameSummary = () => ({
  questionsConnection: {
    aggregate: {
      count: casual.double + 1,
      __typename: 'aggregate',
    },
    __typename: 'questionsConnection',
  },
  categoryQuestionsConnection: {
    aggregate: {
      count: casual.double,
      __typename: 'aggregate',
    },
    __typename: 'categoryQuestionsConnection',
  },
});

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

export {
  LocalStorageMock,
  fakeQuestion,
  fakeUser,
  fakeAdmin,
  fakeCategory,
  fakeCategoryQuestion,
  fakeTeam,
  fakeGame,
  fakeStudyGame,
  fakeGameSummary,
  fakeGameConfig,
};
