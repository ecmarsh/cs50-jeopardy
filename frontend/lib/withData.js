import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { endpoint, endpointProd } from '../config';
import { LOCAL_STATE_QUERY } from '../components/SoundEffect';
import { ANSWERED_QUESTIONS_QUERY } from '../components/GameScreen';
import gql from 'graphql-tag';

// Setup for SSR w/ next and config endpoint
function createClient({ headers }) {
  const typeDefs = gql`
    type StudyQuestion {
      data: Json
    }
  `;
  const client = new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? endpoint : endpointProd,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    typeDefs,
    // cache
    clientState: {
      resolvers: {
        Mutation: {
          toggleSound(_, variables, { cache }) {
            // read the soundOn value from the cache
            const { soundOn } = cache.readQuery({
              query: LOCAL_STATE_QUERY,
            });
            // Toggle sound on || off
            const data = {
              data: { soundOn: !soundOn },
            };
            cache.writeData(data);
            return data;
          },
          addAnswered(_, { id, val }, { cache }) {
            // read the answered questions from the cache
            const { answeredQuestions } = cache.readQuery({
              query: ANSWERED_QUESTIONS_QUERY,
            });
            const newQuestion = { id, val, __typename: 'StudyQuestion' };
            const jsonified = JSON.stringify([
              newQuestion,
              ...answeredQuestions,
            ]);
            // Add new question to data
            const data = {
              data: {
                answeredQuestions: [newQuestion, ...answeredQuestions],
              },
            };

            cache.writeData(data);
            return data;
          },
          resetBoard(_, variables, { cache }) {
            // Reset answered questions
            const data = {
              data: {
                answeredQuestions: [],
              },
            };
            cache.writeData(data);
            return data;
          },
        },
      },
      defaults: {
        soundOn: false,
        answeredQuestions: [],
      },
    },
  });
  return client;
}

export default withApollo(createClient);
