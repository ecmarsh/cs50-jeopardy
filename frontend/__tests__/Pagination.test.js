import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import Router from 'next/router';
import { MockedProvider } from 'react-apollo/test-utils';
import Pagination, { PAGINATION_QUERY } from '../components/Pagination';
import { fakeGame } from '../lib/testUtils';

// Override prefetch error
Router.router = {
  push() {},
  prefetch() {},
};

// Use fakeGame as gamename variable
const { name } = fakeGame();

// Imitate pagination response
function makeMocksFor(length) {
  return [
    {
      request: {
        query: PAGINATION_QUERY,
        variables: {
          gameName: name,
        },
      },
      result: {
        data: {
          categoriesConnection: {
            aggregate: {
              count: length,
              __typename: 'count',
            },
            __typename: 'aggregate',
          },
        },
      },
    },
  ];
}

// Pagination tests
describe('Categories paginated', () => {
  it('displays a loading message', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(1)}>
        <Pagination gameName={name} round={1} />
      </MockedProvider>
    );
    expect(wrapper.text()).toContain('Loading');
  });
  it('renders pagination for 11 categories', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(11)}>
        <Pagination gameName={name} round={1} />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    const pagination = wrapper.find('[data-test="pagination"]');
    expect(toJSON(pagination)).toMatchSnapshot();
  });
  it('renders the round based on paginated categories', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(11)}>
        <Pagination gameName={name} round={3} />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    const roundName = wrapper.find('[data-test="pagination-round"]');
    expect(roundName.text()).toContain('Final Jeopardy');
  });
  it('disables prev button on first round', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(11)}>
        <Pagination gameName={name} round={1} />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(wrapper.find('a.prev').prop('aria-disabled')).toEqual(true);
    expect(wrapper.find('a.next').prop('aria-disabled')).toEqual(false);
  });
  it('disables next button on last round', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(11)}>
        <Pagination gameName={name} round={3} />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(wrapper.find('a.prev').prop('aria-disabled')).toEqual(false);
    expect(wrapper.find('a.next').prop('aria-disabled')).toEqual(true);
  });
  it('enables all buttons on second round', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(11)}>
        <Pagination gameName={name} round={2} />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(wrapper.find('a.prev').prop('aria-disabled')).toEqual(false);
    expect(wrapper.find('a.next').prop('aria-disabled')).toEqual(false);
  });
});
