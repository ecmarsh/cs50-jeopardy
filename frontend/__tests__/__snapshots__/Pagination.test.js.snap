// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories paginated renders pagination for 11 categories 1`] = `
Array [
  <Paginated__PaginationStyled
    data-test="pagination"
  >
    <StyledComponent
      data-test="pagination"
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "Paginated__PaginationStyled-sc-1sq5ena-0",
            "isStatic": false,
            "lastClassName": "cfduEE",
            "rules": Array [
              "display:flex;justify-content:space-around;align-items:baseline;padding:1rem 20%;margin:0 auto;line-height:1;a,a:visited{border:0px solid ",
              [Function],
              ";min-height:100%;font-size:smaller;color:",
              [Function],
              ";&[aria-disabled='true']{visibility:hidden;}span{padding:0 3px;transform:translate3d(0,0,0);display:inline-block;opacity:1;transition:transform 0.2s cubic-bezier(0.04,1.16,0.96,1.03);}}a.prev{order:0;&:hover{span{transform:translate3d(-3px,0,0);}}}a.next{order:2;&:hover{span{transform:translate3d(3px,0,0);}}}h2,h3,h4,h5{order:1;}",
              [Function],
            ],
          },
          "displayName": "Paginated__PaginationStyled",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "Paginated__PaginationStyled-sc-1sq5ena-0",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <div
        className="Paginated__PaginationStyled-sc-1sq5ena-0 cfduEE"
        data-test="pagination"
      >
        <h4
          data-test="pagination-round"
        >
          Single Jeopardy
        </h4>
        <Link
          href={
            Object {
              "pathname": "load",
              "query": Object {
                "game": "Test Game",
                "round": 1,
              },
            }
          }
          prefetch={true}
        >
          <a
            aria-disabled={true}
            className="prev"
            href="load?game=Test%20Game&round=1"
            onClick={[Function]}
          >
            <span>
              ←
            </span>
            Prev Round
          </a>
        </Link>
        <Link
          href={
            Object {
              "pathname": "load",
              "query": Object {
                "game": "Test Game",
                "round": 2,
              },
            }
          }
          prefetch={true}
        >
          <a
            aria-disabled={false}
            className="next"
            href="load?game=Test%20Game&round=2"
            onClick={[Function]}
          >
            Next Round
            <span>
              →
            </span>
          </a>
        </Link>
      </div>
    </StyledComponent>
  </Paginated__PaginationStyled>,
  <StyledComponent
    data-test="pagination"
    forwardedComponent={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "Paginated__PaginationStyled-sc-1sq5ena-0",
          "isStatic": false,
          "lastClassName": "cfduEE",
          "rules": Array [
            "display:flex;justify-content:space-around;align-items:baseline;padding:1rem 20%;margin:0 auto;line-height:1;a,a:visited{border:0px solid ",
            [Function],
            ";min-height:100%;font-size:smaller;color:",
            [Function],
            ";&[aria-disabled='true']{visibility:hidden;}span{padding:0 3px;transform:translate3d(0,0,0);display:inline-block;opacity:1;transition:transform 0.2s cubic-bezier(0.04,1.16,0.96,1.03);}}a.prev{order:0;&:hover{span{transform:translate3d(-3px,0,0);}}}a.next{order:2;&:hover{span{transform:translate3d(3px,0,0);}}}h2,h3,h4,h5{order:1;}",
            [Function],
          ],
        },
        "displayName": "Paginated__PaginationStyled",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "Paginated__PaginationStyled-sc-1sq5ena-0",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      }
    }
    forwardedRef={null}
  >
    <div
      className="Paginated__PaginationStyled-sc-1sq5ena-0 cfduEE"
      data-test="pagination"
    >
      <h4
        data-test="pagination-round"
      >
        Single Jeopardy
      </h4>
      <Link
        href={
          Object {
            "pathname": "load",
            "query": Object {
              "game": "Test Game",
              "round": 1,
            },
          }
        }
        prefetch={true}
      >
        <a
          aria-disabled={true}
          className="prev"
          href="load?game=Test%20Game&round=1"
          onClick={[Function]}
        >
          <span>
            ←
          </span>
          Prev Round
        </a>
      </Link>
      <Link
        href={
          Object {
            "pathname": "load",
            "query": Object {
              "game": "Test Game",
              "round": 2,
            },
          }
        }
        prefetch={true}
      >
        <a
          aria-disabled={false}
          className="next"
          href="load?game=Test%20Game&round=2"
          onClick={[Function]}
        >
          Next Round
          <span>
            →
          </span>
        </a>
      </Link>
    </div>
  </StyledComponent>,
  <div
    className="Paginated__PaginationStyled-sc-1sq5ena-0 cfduEE"
    data-test="pagination"
  >
    <h4
      data-test="pagination-round"
    >
      Single Jeopardy
    </h4>
    <Link
      href={
        Object {
          "pathname": "load",
          "query": Object {
            "game": "Test Game",
            "round": 1,
          },
        }
      }
      prefetch={true}
    >
      <a
        aria-disabled={true}
        className="prev"
        href="load?game=Test%20Game&round=1"
        onClick={[Function]}
      >
        <span>
          ←
        </span>
        Prev Round
      </a>
    </Link>
    <Link
      href={
        Object {
          "pathname": "load",
          "query": Object {
            "game": "Test Game",
            "round": 2,
          },
        }
      }
      prefetch={true}
    >
      <a
        aria-disabled={false}
        className="next"
        href="load?game=Test%20Game&round=2"
        onClick={[Function]}
      >
        Next Round
        <span>
          →
        </span>
      </a>
    </Link>
  </div>,
]
`;
